kong/db/dao/plugins/external.lua

```diff
@@ -0,0 +1,642 @@
+      plugin[phase] = function(self, conf)
+        local saved = {
+          serialize_data = kong.log.serialize(),
+          ngx_ctx = ngx.ctx,
+          ctx_shared = kong.ctx.shared,
+        }
+
+        ngx_timer_at(0, function()
+          local co = coroutine.running()
+          save_for_later[co] = saved
+
+          handle_event(server_rpc, self.name, conf, phase)
+
+          save_for_later[co] = nil
+        end)
+      end
+
+    else
+      plugin[phase] = function(self, conf)
+        handle_event(server_rpc, self.name, conf, phase)
+      end
+    end
+  end
+
+  return plugin
+end
+
+
+--[[
+
+Plugin info requests
+
+Disclaimer:  The best way to do it is to have "ListPlugins()" and "GetInfo(plugin)"
+RPC methods; but Kong would like to have all the plugin schemas at initialization time,
+before full cosocket is available.  At one time, we used blocking I/O to do RPC at
+non-yielding phases, but was considered dangerous.  The alternative is to use
+`io.popen(cmd)` to ask fot that info.
+
+
+In the external plugins configuration, the `.info_cmd` field contains a string
+to be executed as a command line.  The output should be a JSON string that decodes
+as an array of objects, each defining the name, priority, version and schema of one
+plugin.
+
+    [{
+      "name": ... ,
+      "priority": ... ,
+      "version": ... ,
+      "schema": ... ,
+      "phases": [ phase_names ... ],
+    },
+    {
+      ...
+    },
+    ...
+    ]
+
+This array should describe all plugins currently available through this server,
+no matter if actually enabled in Kong's configuration or not.
+
+--]]
+
+
+local function register_plugin_info(server_def, plugin_info)
+  if _plugin_infos[plugin_info.Name] then
+    kong.log.err(string.format("Duplicate plugin name [%s] by %s and %s",
+      plugin_info.Name, _plugin_infos[plugin_info.Name].server_def.name, server_def.name))
+    return
+  end
+
+  _plugin_infos[plugin_info.Name] = {
+    server_def = server_def,
+    --rpc = server_def.rpc,
+    name = plugin_info.Name,
+    PRIORITY = plugin_info.Priority,
+    VERSION = plugin_info.Version,
+    schema = plugin_info.Schema,
+    phases = plugin_info.Phases,
+  }
+end
+
+local function ask_info(server_def)
+  if not server_def.query_command then
+    kong.log.info(string.format("No info query for %s", server_def.name))
+    return
+  end
+
+  local fd, err = io.popen(server_def.query_command)
+  if not fd then
+    local msg = string.format("loading plugins info from [%s]:\n", server_def.name)
+    kong.log.err(msg, err)
+    return
+  end
+
+  local infos_dump = fd:read("*a")
+  fd:close()
+  local infos = cjson_decode(infos_dump)
+  if type(infos) ~= "table" then
+    error(string.format("Not a plugin info table: \n%s\n%s",
+        server_def.query_command, infos_dump))
+    return
+  end
+
+  for _, plugin_info in ipairs(infos) do
+    register_plugin_info(server_def, plugin_info)
+  end
+end
+
+local function load_all_infos()
```

Comment:
Since it is used only once, I suggest renaming this to `get_plugin_info(plugin_name)` and deal with the memoization of `_plugin_infos` completely inside this function

Reply:
done